### PR TITLE
CORDA-2965 Make Tx verification exceptions serializable.

### PR DESCRIPTION
**Pull request by tudor-malene**
see https://r3-cev.atlassian.net/browse/CORDA-2965

Note: the changes to the detekt config are required to ignore the gradle `buildSrc` folder which is created sometimes.

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/contracts/TransactionVerificationException.kt [hunk 1]
```diff
@@ -207,12 +209,11 @@ abstract class TransactionVerificationException(val txId: SecureHash, message: S
             For details see: https://docs.corda.net/api-contract-constraints.html#contract-state-agreement
             """.trimIndent(), null)
 
-
     /**
      * If the network parameters associated with an input or reference state in a transaction are more recent than the network parameters of the new transaction itself.
      */
     @KeepForDJVM
-    class TransactionNetworkParameterOrderingException(txId: SecureHash, inputStateRef: StateRef, txnNetworkParameters: NetworkParameters, inputNetworkParameters: NetworkParameters)
+    class TransactionNetworkParameterOrderingException(txId: SecureHash, val inputStateRef: StateRef, val txnNetworkParameters: NetworkParameters, val inputNetworkParameters: NetworkParameters)
```
Comment (r3domfox) [inline core/src/main/kotlin/net/corda/core/contracts/TransactionVerificationException.kt]: It may be better to follow the practice of some of the other exceptions, and have the primary constructor take a `String` message, along with the `txId`, and provide a secondary constructor which takes the network parameters, etc., and constructs the message. We do not really need to send these objects over the wire - we just need to provide a meaningful error message.

Comment (tudor-malene) [reply core/src/main/kotlin/net/corda/core/contracts/TransactionVerificationException.kt]: This is part of the public API now.
Unfortunately, we have to keep all these properties

Comment (r3domfox) [reply core/src/main/kotlin/net/corda/core/contracts/TransactionVerificationException.kt]: Yes, but I think we can move them to a secondary constructor. We did so with some of the others, IIRC.

Comment (r3domfox) [reply core/src/main/kotlin/net/corda/core/contracts/TransactionVerificationException.kt]: In terms of Java API compatibility, I think there's no impact. If the API checker fusses, we can correct it.

File: core-tests/src/test/kotlin/net/corda/coretests/contracts/TransactionVerificationExceptionSerialisationTests.kt [hunk 2]
```diff
@@ -23,7 +33,11 @@ class TransactionVerificationExceptionSerialisationTests {
     private fun defaultFactory() = SerializerFactoryBuilder.build(
             AllWhitelist,
             ClassLoader.getSystemClassLoader()
-    ).apply { register(ThrowableSerializer(this)) }
+    ).apply {
+        register(ThrowableSerializer(this))
+        register(DurationSerializer(this))
+        register(InstantSerializer(this))
```
Comment (r3domfox) [inline core-tests/src/test/kotlin/net/corda/coretests/contracts/TransactionVerificationExceptionSerialisationTests.kt]: Do you now need to register these extra serializers?

File: detekt-config.yml [hunk 3]
```diff
@@ -10,29 +10,33 @@ complexity:
...
     threshold: 4
   ComplexMethod:
     active: true
+    excludes: "**/buildSrc/**"
     threshold: 10
     ignoreSingleWhenExpression: true
   LargeClass:
     active: true
-    excludes: "**/test/**,**/integration-test/**,**/integration-test-slow/**,**/*Test.kt,**/*Tests.kt"
+    excludes: "**/test/**,**/integration-test/**,**/integration-test-slow/**,**/*Test.kt,**/*Tests.kt,**/buildSrc/**"
```
Comment (r3domfox) [inline detekt-config.yml:22]: Are these changes intended for this PR, or do you need to rebase?

**Review by r3domfox — COMMENTED** [review]
I think this is the correct approach, but should mean that the additional custom serializers no longer need to be registered.

File: core-tests/src/test/kotlin/net/corda/coretests/contracts/TransactionVerificationExceptionSerialisationTests.kt (hunk 2 above)
Comment (tudor-malene) [reply core-tests/src/test/kotlin/net/corda/coretests/contracts/TransactionVerificationExceptionSerialisationTests.kt]: good point. forgot about 'em

File: detekt-config.yml (hunk 3 above)
Comment (tudor-malene) [reply detekt-config.yml:22]: It is required for anyone who has installed some plugin over the time that installs stuff in the `buildSrc` folder of gradle.
If I don't exclude them, when I run detekt it will show a lot of irrelevant errors

Comment (tudor-malene) [reply detekt-config.yml:22]: I could split it up in a separate PR if you think

**Review by r3domfox — APPROVED** [review]
LGTM
